tfhe/src/strings/char_iter.rs

```diff
@@ -0,0 +1,211 @@
+use super::ciphertext::FheAsciiChar;
+use rayon::iter::{IntoParallelIterator, IntoParallelRefIterator, ParallelIterator};
+
+pub(super) type CharIter<'a> = OptionalEndSliceIter<'a, FheAsciiChar>;
+
+pub(super) struct OptionalEndSliceIter<'a, T> {
+    slice: &'a [T],
+    last: Option<&'a T>,
+}
+
+impl<'a, T> Clone for OptionalEndSliceIter<'a, T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<'a, T> Copy for OptionalEndSliceIter<'a, T> {}
+
+impl<'a, T> OptionalEndSliceIter<'a, T> {
+    pub(super) fn len(&self) -> usize {
+        self.slice.len() + if self.last.is_some() { 1 } else { 0 }
+    }
+
+    pub(super) fn new(slice: &'a [T], last: Option<&'a T>) -> Self {
+        Self { slice, last }
+    }
+}
+
+pub mod iter {
+    use super::*;
+    use std::cmp::Ordering;
+
+    impl<'a, T> IntoIterator for OptionalEndSliceIter<'a, T> {
+        type Item = &'a T;
+
+        type IntoIter = OptionalEndSliceIterator<'a, T>;
+
+        fn into_iter(self) -> Self::IntoIter {
+            OptionalEndSliceIterator {
+                iter: self,
+                index: 0,
+                back_index: self.len(),
+            }
+        }
+    }
+
+    pub struct OptionalEndSliceIterator<'a, T> {
+        iter: OptionalEndSliceIter<'a, T>,
+        index: usize,
+        back_index: usize,
```

Comment:
Would it be easier to resuse std::slice::Iter<'a> make it easier ?

```rust


struct OptionalEndSliceIter<'a, T> {
    slice_iter: std::slice::Iter<'a, T>,
    last: Option<&'a T>,
}

impl<'a, T> OptionalEndSliceIter<'a, T> {
    pub fn len(&self) -> usize {
        self.slice_iter.as_slice().len() + usize::from(self.last.is_some())
    }

    pub fn new(slice: &'a [T], last: Option<&'a T>) -> Self {
        Self { slice_iter: slice.iter(), last }
    }
}

impl<'a, T> Iterator for OptionalEndSliceIter<'a, T> {
    type Item = &'a T;
    
    fn next(&mut self) -> Option<Self::Item> {
        let thing = self.slice_iter.next();
        if thing.is_some() {
            return thing;
        }
        
        self.last.take()
    }
}


impl<'a, T> DoubleEndedIterator for OptionalEndSliceIter<'a, T> {
    fn next_back(&mut self) -> Option<Self::Item> {
         let thing = self.slice_iter.next_back();
        if thing.is_some() {
            return thing;
        }
        
        self.last.take()
    }
}
```

Reply:
Sure, thanks for the idea